internal/ext/unsafex/unsafex.go

```diff
@@ -17,42 +17,124 @@
+	~int8 | ~int16 | ~int32 | ~int64 | ~int |
+		~uint8 | ~uint16 | ~uint32 | ~uint64 | ~uint |
+		~uintptr
+}
+
+// Layout is the layout of a type.
+//
+// This is a more convenient abstraction that manipulating the size and
+// alignment separately.
+type Layout struct {
+	Size, Align int
+
+	// TODO: Add something like PointerFree? It's not clear if Go even exposes
+	// an easy way to compute this. There's certainly no intrinsic for it...
 }
 
-// Index is like [unsafe.Add], but it operates on a typed pointer and scales the
+// LayoutOf returns the layout of some type.
+func LayoutOf[T any]() Layout {
+	var v T
+	return Layout{
+		Size:  int(unsafe.Sizeof(v)),
+		Align: int(unsafe.Alignof(v)),
+	}
+}
+
+// Add is like [unsafe.Add], but it operates on a typed pointer and scales the
 // offset by that type's size, similar to pointer arithmetic in Rust or C.
 //
 // This function has the same safety caveats as [unsafe.Add].
-func Index[T any](p *T, idx int) *T {
+//
+//go:nosplit
+func Add[P ~*E, E any, I Int](p P, idx I) P {
 	raw := unsafe.Pointer(p)
-	raw = unsafe.Add(raw, idx*Size[T]())
-	return (*T)(raw)
+	raw = unsafe.Add(raw, int(idx)*LayoutOf[E]().Size)
+	return P(raw)
 }
 
-// Cast performs an unchecked cast of one pointer type to another, through
-// an [unsafe.Pointer].
-func Cast[To, From any](p *From) *To {
-	return (*To)(unsafe.Pointer(p))
-}
-
-// Transmute bit-casts a value of type From to a value of type To.
+// Bitcast bit-casts a value of type From to a value of type To.
 //
 // This operation is very dangerous, because it can be used to break package
 // export barriers, read uninitialized memory, and forge pointers in violation
 // of [unsafe.Pointer]'s contract, resulting in memory errors in the GC.
 //
-// Transmute should only be used on types the caller owns, and those types
-// should have identical GC shapes.
-func Transmute[To, From any](p From) To {
-	return *Cast[To](&p)
+// Panics if To and From have different sizes.
+//
+//go:nosplit
+func Bitcast[To, From any](v From) To {
+	// This function is correctly compiled down to a mov, as seen here:
+	// https://godbolt.org/z/qvndcYYba
+	//
+	// With redundant code removed, stenciling Bitcast[float64, int64] produces
+	// (as seen in the above Godbolt):
+	//
+	//   TEXT    unsafex.Bitcast[float64,int64]
+	//   MOVQ    32(R14), R12
+	//   TESTQ   R12, R12
+	//   JNE     morestack
+	//   XCHGL   AX, AX
+	//   MOVQ    AX, X0
+	//   RET
+
+	if LayoutOf[To]().Size != LayoutOf[From]().Size {
+		// This check will always be inlined away, because Bitcast is
+		// manifestly inline-able.
+		//
+		// NOTE: This could potentially be replaced with a link error, by making
+		// this call a function with no body (and then not defining that
+		// function in a .s file; although, note we do need an empty.s to
+		// silence a compiler error in that case).
+		panic(badBitcast[To, From]{})
+	}
+
+	// To avoid an unaligned load below, we copy From into
+	// a struct aligned to To.
```

Comment:
Why is this necessary? Is it not already the case that `v` must be aligned in memory (on the stack)? What about if you instead did `&v` -- would that not ensure that the value is already aligned without these shenanigans?

Reply:
It is insufficient. I have added a comment explaining why this is necessary.